### PR TITLE
HexEditor: Add strings to the value inspector

### PR DESCRIPTION
**Pull request by ebanner**
Implements https://github.com/SerenityOS/serenity/issues/13655

Continuing work done by @MetallicSquid and @hamburger1984  in https://github.com/SerenityOS/serenity/pull/13809 and https://github.com/SerenityOS/serenity/pull/15590

image

### CODE REVIEW

**Comment by ebanner** [comment]
@kleinesfilmroellchen @hamburger1984 @MetallicSquid wondering if I could get your guy's feedback. Main open question for me is what we do with displaying control characters... should we replace them with `SPC`, or some custom rendering (e.g. `<nl>` for newlines)...?

**Comment by kleinesfilmroellchen** [comment]
> Hey @kleinesfilmroellchen @hamburger1984 @MetallicSquid 👋
> 
> Wondering if I could get your guys' feedback
> 
> Main open question for me is what we do with displaying control characters... should we replace them with `' '`, or some other custom rendering (e.g. `<nl>` for newlines)...?
> 
> I thought that perhaps this issue could be handled at the LibGUI layer, but thinking about it some more, we probably do want newlines to be rendered as newlines...

Using \n, \r, etc. as replacements is probably best because many will be familiar with the notation. You should do that just in HexEditor and not in LibGUI.

Also, please clean up your commits and if you can, use the previous PR's commits as-is with your changes as commit amends (if they just change what has already been done, not add anything more)

**Comment by ebanner** [comment]
> Using \n, \r, etc. as replacements is probably best because many will be familiar with the notation. You should do that just in HexEditor and not in LibGUI.

👍

> Also, please clean up your commits and if you can, use the previous PR's commits as-is with your changes as commit amends (if they just change what has already been done, not add anything more)

Sounds good. In order to get the previous commits, it seems like I'll have to create a new branch off @hamburger1984's branch. The only other thing I can think to do is to rebase (?) the commits from https://github.com/SerenityOS/serenity/pull/15590 onto this branch. However, this branch is like 3.5k commits ahead of https://github.com/SerenityOS/serenity/pull/15590 and I really don't know enough about rebasing to reconcile the commit history difference.

I'll go off and create a new branch and new PR. I'll leave this branch up in case someone with rebasing expertise wants to propose some way to bring those commits over to this branch.

**Comment by ebanner** [comment]
Yeah, sure. That would save me the extra step of fast forwarding (?) your branch up to date with master.

However, if you aren't able to do right away, don't worry about it. I think fast forwarding the new branch I base of yours is like a one-command simple thing, and I don't want to make it an unnecessary blocker.

**Comment by ebanner** [comment]
Actually, maybe! If you fast forward your branch, I might be able to rebase it onto this branch.

**Comment by hamburger1984** [comment]
I rebased my branch onto the current master. Hope this doesn't mess up your changes ;-)

**Comment by ebanner** [comment]
Oh you're good. I don't see how it would.

So that I can do the same on my end (before I rebase (?) your branch onto mine [?]), what command did you run to rebase your branch off master?

**Comment by hamburger1984** [comment]
I added your commit onto my fork.. maybe you can just update your branch to that?

What I did might not make too much sense in your fork. My names or remotes might be totally different from yours..

**Comment by ebanner** [comment]
Ok, I pulled in the commits from the other PR, with the help of @hamburger1984 🙌

I think the last thing we should do is rewrite the commit history so that it usefully communicates what the changes were made

For example, we should do something with (delete?) bdac3234c97d5721a0eafc5b31956a7a7ec61a2b because it says it adds UTF16 to the inspector view, but it does not 

I propose condensing it down to just 2 commits

```
- Add ascii and utf8 to the value inspector
- LibTextCoded refactor and add UTF16 view
```

We could expand it out into several commits like

```
- Add ASCII to the value inspector
- Add UTF8 to the value inspector
- Add UTF16 view to the value inspector
- Refactor decoding to use LibTextCoded
- Display control characters in value inspector
```

What do you guys think?

**Comment by ebanner** [comment]
Ok, so I felt kinda weird about basically undoing work from @MetallicSquid and @hamburger1984, so I left your guys' implementations of ASCII string & UTF-8 string in there (maybe we can refactor at a later date if anyone cares to)

Commit history is looking good to me. Only thing is now we don't have record of @kleinesfilmroellchen contributing (let me know if you want to get in here somehow @kleinesfilmroellchen)

Otherwise I'm marking this PR as ready to review. I think it's ready to merge

**Comment by ebanner** [comment]
@kleinesfilmroellchen do you have merge privileges? Do we need to find someone else?

**Comment by kleinesfilmroellchen** [comment]
@ebanner I'm not a maintainer, and I also need to have a look if this is now in a final state.

**Comment by ebanner** [comment]
It's in the final state 👍

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (kleinesfilmroellchen) [inline Userland/Applications/HexEditor/HexEditorWidget.cpp:302]: auto

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (kleinesfilmroellchen) [inline Userland/Applications/HexEditor/HexEditorWidget.cpp]: This commit should probably be squashed with the next one

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (kleinesfilmroellchen) [inline Userland/Applications/HexEditor/HexEditorWidget.cpp]: What happens here if the UTF-16 is not valid?

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (kleinesfilmroellchen) [inline Userland/Applications/HexEditor/HexEditorWidget.cpp:386]: This will append valid UTF-8, which we don't want since it's supposed to be an ASCII string. The control character replacement will not fix this.

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (kleinesfilmroellchen) [inline Userland/Applications/HexEditor/HexEditorWidget.cpp:620]: Missing here are at least bell \a, and most <64 ASCII control characters.

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (ebanner) [reply Userland/Applications/HexEditor/HexEditorWidget.cpp]: 👌

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (ebanner) [reply Userland/Applications/HexEditor/HexEditorWidget.cpp]: Hmm I think we're good here, because we're going byte by byte

At the single byte level, ASCII == UTF-8. If we append a byte that has value >127, then that's an invalid ASCII _as well as_ UTF-8 character, and will result in a `<?>` glyph, which is the desired behavior here

Let me know if you had something else in mind

Comment (ebanner) [reply Userland/Applications/HexEditor/HexEditorWidget.cpp]: Hmm I think we're good here, because we're going byte by byte

At the single byte level, ASCII == UTF-8. If we append a byte that has value >127, then that's an invalid ASCII _as well as_ UTF-8 character, and will result in a `<?>` glyph, which is the desired behavior

Let me know if you had something else in mind

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (ebanner) [reply Userland/Applications/HexEditor/HexEditorWidget.cpp:620]: This is true. I actually chose the characters I in `make_display_string` by seeing which ones were displayed literally. All the ones I left out are rendered with `<?>`, which I was fine with

I see now (after responding to your above comment), the interpretation of `<?>` is ambiguous. Up until this point, I was assuming it could either mean "invalid character" or "some ascii control character"

My goal with `make_display_string` was to narrowly fix the issue of having it be so control characters obfuscate later characters (like `\n`)

However, I realize now, the way we render various control characters could change tomorrow (they could start rendering not as `<?>`) and we would have to fix it, so it does seem like we should handle all the control characters

I'll have to think whether I'm willing to do that in this PR, or whether it's out of scope

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (ebanner) [reply Userland/Applications/HexEditor/HexEditorWidget.cpp]: Hmm it seems like it gives me garbled/garbage characters. But I haven't studied UTF-16 well enough to verify that these characters are indeed garbage and not valid

This seems like another issue. What happens if they are invalid? Seems like we should show `<?>` glyphs for each invalid character. There are weird edge cases when we select, say, a single byte. Should we show a `<?>` character, or nothing at all...? 🤔

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (kleinesfilmroellchen) [reply Userland/Applications/HexEditor/HexEditorWidget.cpp:620]: \a is not affected by this question so it should be included either way. There are proper replacement characters defined by Unicode in some higher code points which are supposed to be used for this purpose. However, if Serenity has a sane way of rendering them at the moment, I guess that's fine.

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (kleinesfilmroellchen) [reply Userland/Applications/HexEditor/HexEditorWidget.cpp]: > Hmm it seems like it gives me garbled/garbage characters. Though I haven't studied UTF-16 well enough to verify that these characters are indeed garbage or whether they're actually valid

Right, it is much safer to show an "invalid" character or some sort of meta "invalid" message instead of garbage data.

> Another issue is what happens if they are invalid? Seems like we should show `<?>` glyphs for each invalid character, not a garbled character

If a single UTF-16 two-byte sequence is invalid, we should ideally show some replacement character which signals "invalid encoding" to the user instead of just "valid encoding for absent code point". However we have no useful way of doing that right now, so I guess � is fine for now. We should definitely not invalidate the entire string if possible, as that is not useful information; imagine the user trying to figure out which part of a long string in a binary file is invalid.

> There are weird edge cases though, when we select, say, a single byte. Should we show a `<?>` character, or no character at all...? 🤔

A single byte is never valid UTF-16, so we can show nothing at all. It's the same as with too few bytes for the multi-byte number types.

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (kleinesfilmroellchen) [reply Userland/Applications/HexEditor/HexEditorWidget.cpp]: > At the single byte level, ASCII == UTF-8. If we append a byte (i.e. char) that has value >127, then that's an invalid ASCII _as well as_ UTF-8 character, and will result in a `<?>` glyph, which is the desired behavior
>
> I think we would only run into the issue you lay out here if we were appending a _code point_

The problem, as you correctly identify, is with non-ASCII multibyte sequences. In UTF-8, many of those are valid single (!) codepoints, while in ASCII, they are many invalid characters. This is an important distinction to make if the format the user is working with is explicitly requiring valid ASCII and e.g. the user tries to figure out why some string is not accepted by a reader program.

I think it is very important to be strict here. A field that says "ASCII" should only parse and display valid ASCII as defined by ANSI and ISO many decades ago. Because many (especially old) parsers will be strict about this and even if we are strict, the user can still decide to ignore invalid ASCII characters if need be.

Comment (ebanner) [reply Userland/Applications/HexEditor/HexEditorWidget.cpp]: Sorry to say, but I _really_ don't grok what you are saying here

If you could give an example, that would go a really long way to help me understand the point you are getting at

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (ebanner) [reply Userland/Applications/HexEditor/HexEditorWidget.cpp]: > Right, it is much safer to show an "invalid" character or some sort of meta "invalid" message instead of garbage data.

👍

Going back to above, I take back what I said about garbled characters. I cannot (in fact) verify that garbled data is produced. That is because the only code points in the UTF-16 2 byte range are `U+D800` to `U+DFFF` , and these decode to `<?>` (expected behavior). Any other "weird" characters I get by randomly typing in 2 bytes pairs I assume are, in fact, valid UTF-16 decoded characters (until I see otherwise)

> absent code point

By “absent code point”, do you mean “code point corresponding to a character that serenity doesn’t know how to display”?

> imagine the user trying to figure out which part of a long string in a binary file is invalid

So it’s kinda difficult… the UTF-16 decoder is resilient when I insert an invalid (2 byte) character in the middle of a string of (2 byte) characters (e.g `$$<?>$`). However, if I append a null byte in front of a sequence of valid UTF-16 characters, the whole thing is decoded incorrectly

Seems like this is getting into the general problem of decoding UTF-16, what’s valid, what’s not, which seems really hard. I’d prefer to leave that logic in the UTF-16 decoder for now, and wait until we have some issues and take those case by case to see if we should deviate from the UTF-16 decoder and something custom in HexEditor

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (ebanner) [reply Userland/Applications/HexEditor/HexEditorWidget.cpp]: So, I wound up removing display string logic. Reason being, I don't have a good sense for what we should actually do

I'm considering it outside the scope of this issue. I plan to create a new issue and create another PR that references it so if someone with a better sense can pick it up and run with it

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (ebanner) [reply Userland/Applications/HexEditor/HexEditorWidget.cpp:402]: > A single byte is never valid UTF-16, so we can show nothing at all. It's the same as with too few bytes for the multi-byte number types.

Following up, I confirmed that nothing is diplayed in the UTF-16 String view when a single byte is highlighted

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (ebanner) [reply Userland/Applications/HexEditor/HexEditorWidget.cpp]: The claim I believe you are making is that there can be a byte that is not valid ASCII that will be shown in the ASCII display view with the current logic

If that is correct, please give me an example of such a byte that will produce the bug. If that's not correct, please do clarify what you mean

**Comment by ebanner** [comment]
@kleinesfilmroellchen just FYI heads up, I'm currently waiting for your feedback/approval to proceed further

**Comment by ebanner** [comment]
Shoot, I resolved the merge conflict within github and no I don't know how to squash that latest commit (rebase isn't working like I expect) 🙈

**Comment by ebanner** [comment]
Fixed the merge conflict, cleaned up the commit history. Ready for feedback @kleinesfilmroellchen!

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (kleinesfilmroellchen) [inline Userland/Applications/HexEditor/HexEditorWidget.cpp]: Still, auto.

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (kleinesfilmroellchen) [inline Userland/Applications/HexEditor/HexEditorWidget.cpp]: no curlies

**Review by kleinesfilmroellchen — APPROVED** [review]
If you have a minute please clean up the commit history (I know the code is immediately removed in the next commit), other than that LGTM

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (ebanner) [reply Userland/Applications/HexEditor/HexEditorWidget.cpp]: I don't know what's going on, but this comment looks to be pointing to some old commit. GitHub is confused. Check the "files changed" tab and you'll see it's `auto`

Resolving this comment due to the misleading nature of it

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (ebanner) [reply Userland/Applications/HexEditor/HexEditorWidget.cpp]: Oh, cleaning up the git commit history... I see what you are saying. Yeah, let me look into that...

**Comment by ebanner** [comment]
Oh no... hamburger's commit got lost in the last commit. I'll look into getting that back

**Comment by ebanner** [comment]
OK so I condensed everything into a single commit, thereby killing two birds with one stone

The only reason I left it at two commits is to better show who did what work. But if preserving the two commits means I have to modify the code after the first commit to be "production ready"... I'm not willing to do that

So all that's left is I'm gonna spot check run everything one more time to see if it still works. Let me know if the single commit message could use some tweaking

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (ebanner) [reply Userland/Applications/HexEditor/HexEditorWidget.cpp]: I condensed both commits together, so this issue is now defunct/superfluous

**Review by kleinesfilmroellchen — APPROVED** [review]

File: Userland/Applications/HexEditor/HexEditor.cpp
Comment (AtkinsSJ) [inline Userland/Applications/HexEditor/HexEditor.cpp:250]: Semantically, a `ByteBuffer` is more what we want to use here.

Also, you know how large you want it to be, so you can call `ensure_capacity()` on the vector/bytebuffer which will allocate once instead of growing the buffer repeatedly.

File: Userland/Applications/HexEditor/HexEditorWidget.cpp
Comment (AtkinsSJ) [inline Userland/Applications/HexEditor/HexEditorWidget.cpp:389]: This StringBuilder isn't needed, you can instead treat `selected_bytes`'s data as a DeprecatedString directly:

```suggestion
    value_inspector_model->set_parsed_value(ValueInspectorModel::ValueType::ASCIIString, DeprecatedString { selected_bytes });
```

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
I have a couple of suggestions. Sorry to again postpone this getting merged! :sweat_smile: 

As a side-note, it's weird how HexEditor assumes big-endian UTF16 doesn't exist.

File: Userland/Applications/HexEditor/HexEditor.cpp
Comment (ebanner) [reply Userland/Applications/HexEditor/HexEditor.cpp]: Got it, makes sense!

File: Userland/Applications/HexEditor/HexEditor.cpp
Comment (ebanner) [inline Userland/Applications/HexEditor/HexEditor.cpp:251]: `ByteBuffer` change

File: Userland/Applications/HexEditor/HexEditor.cpp
Comment (ebanner) [inline Userland/Applications/HexEditor/HexEditor.cpp:252]: `ByteBuffer` change

**Review by AtkinsSJ — APPROVED** [review]

File: Userland/Applications/HexEditor/ValueInspectorModel.h
Comment (ebanner) [reply Userland/Applications/HexEditor/ValueInspectorModel.h:106]: Hi @konradekk 👋

> not sure if this is really relevant, i guess iʼd go for merging ××× and ×××String values (through case fall-through).

If I understand you correctly, yes I can see how that could work. However, I like the fact that the code explicitly mirrors the view inspector in the UI, like there's a 1-to-1 mapping between each line here and a line in the GUI

> while visiting: maybe we could return “Unknown” for non-matching (i.e. default) case? 🤔

Hmm, the feeling I get from your comment is that `type` (the variable being matched on) is like some dynamic value that we're pattern matching against. Whereas my impression of what's happening here is this is like the definition in code for how the UI looks. Doesn't really seem like we need to even "catch" the default case. Let me know if I've got that wrong

Ah, I do see that there is a `default` option there. And if you're thinking `"Unknown"` would be a better value than `""` there, then... I don't really have an opinion one way or another (haven't thought about where this may even come into play - it's not clear to me we have ever or will really ever hit this).

**Comment by ebanner** [comment]
Hey @AtkinsSJ 👋

> As a side-note, it's weird how HexEditor assumes big-endian UTF16 doesn't exist

Yeah, I didn't really notice this before you mentioned it. I don't really have context for why the value inspector is little-endian centric (I'm not sure but I assume other types in the value inspector are all little endian-based). I kinda assumed that SerenityOS is little-endian based and this is why...? But I don't even know that to be the case.

If someone wants big endian UTF-16, we can revisit at that time. Probably making a special case UTF-16BE in the value inspector would be fine for that. If we get more and more requests of big endian stuff, maybe creating separate BE and LE value inspectors would be the way to go

**Comment by ebanner** [comment]
As per an offline discussion with AtkinsSJ, there is big-endian/little-endian setting (that should dictate the endianness of the UTF-16 encoding)

image

Planning on creating an issue off this

File: Userland/Applications/HexEditor/ValueInspectorModel.h
Comment (konradekk) [reply Userland/Applications/HexEditor/ValueInspectorModel.h:106]: oh! i guess i donʼt see the reason for differentiating `×××` and `×××String` variants then! 😅

definitely out of scope but maybe a yak to shave, even for you, as it seems you still plan to work around that code! 😁

Comment (ebanner) [reply Userland/Applications/HexEditor/ValueInspectorModel.h:106]: Just want to make sure I'm understand you

> oh! i guess i donʼt see the reason for differentiating ××× and ×××String variants then! 😅

So, the xxx and xxxString are two different rows in the value inspector. xxx shows the decoded byte (or word) under cursor, whereas xxxString shows the decoded highlighted area

Do you share that understanding, or are you getting at something else?

**Pull request merged**
